gitoid/src/named_digest.rs

```diff
@@ -0,0 +1,35 @@
+//! Trait specifying valid [`GitOid`] hash algorithms.
+
+use crate::named_digest::private::Sealed;
+#[cfg(doc)]
+use crate::GitOid;
+use digest::Digest;
+use sha1::Sha1;
+use sha2::Sha256;
+
+mod private {
+    pub trait Sealed {}
+}
+
+/// Hash algorithms that can be used to make a [`GitOid`].
+///
+/// This is a sealed trait to ensure it's only used for hash
+/// algorithms which are actually supported by Git. No other
+/// types, even if they implement [`Digest`] can implement
+/// this trait.
+pub trait NamedDigest: Digest + Sealed {
+    /// The name of the hash algorithm in lowercase ASCII.
+    const NAME: &'static str;
+}
```

Comment:
This list of impls currently misses SHA-1DC, which is what Git usually uses by default today. I haven't investigated the availability of Rust implementations of this.